File: 2012 Book Source Code/Chapter_10/Chapter 10/Chapter 10/SSRS Viewer RVC VB/GetParameters.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports SSRS_Viewer_RVC_VB.SSRSWebSerivce
Imports System.Web.Services.Protocols
Imports System.Collections


Public Class GetParameters

    Private url As String
    Private server As String
    Private report As String
    Private parameters() As Microsoft.Reporting.WinForms.ReportParameter '= Nothing
    Dim rs As ReportingService2010

    Public Sub GetParameters(ByVal repurl As String)
        'Change sub from New to GetParameters as C# code does
        InitializeComponent()

        url = repurl
        Dim reportInfo() As String = url.Split("?")
        server = reportInfo(0)
        report = reportInfo(1)

    End Sub


    'Public ReadOnly Property Parameters() As Microsoft.Reporting.WinForms.ReportParameter()
    '    Get
    '        Return parameters
    '    End Get
    'End Property


    'Private Sub GetParameters_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

    '    rs = New ReportingService2010()
    '    rs.Credentials = System.Net.CredentialCache.DefaultCredentials

    '    Dim forRendering As Boolean = True
    '    Dim historyID As String = vbNull
    '    Dim values(1) As ParameterValue
    '    Dim credentials() As DataSourceCredentials = Nothing
    '    Dim parametersSSRS() As ItemParameter = Nothing
    '    Dim pvs() As ValidValue = Nothing

    '    Dim x As Int32 = 5
    '    Dim y As Int32 = 30

    '    parametersSSRS = rs.GetItemParameters(report, historyID, forRendering, values, credentials)

    '    Try
    '        values(0) = New ParameterValue()
    '        values(0).Label = "ServiceYear"
    '        values(0).Name = "ServiceYear"
    '        values(0).Value = "2007"

    '        parametersSSRS = rs.GetItemParameters(report, historyID, forRendering, values, credentials)

    '        If parameters IsNot Nothing Then
    '            For Each rp As ItemParameter In parametersSSRS
    '                Me.SuspendLayout()
    '                Me.parameterPanel.SuspendLayout()
    '                Me.parameterPanel.SendToBack()

    '                '' now create a label for the combo box below
    '                Dim lbl As Label = New Label()
    '                lbl.Anchor = (System.Windows.Forms.AnchorStyles.Top)
    '                lbl.Location = New System.Drawing.Point(x, y)
    '                lbl.Name = rp.Name
    '                lbl.Text = rp.Name
    '                lbl.Size = New System.Drawing.Size(150, 20)
    '                Me.parameterPanel.Controls.Add(lbl)
    '                x = x + 150

    '                '' now make a combo box and fill it
    '                Dim a As ComboBox = New ComboBox()
    '                a.Anchor = (System.Windows.Forms.AnchorStyles.Top)
    '                a.Location = New System.Drawing.Point(x, y)
    '                a.Name = rp.Name
    '                a.Size = New System.Drawing.Size(200, 20)
    '                x = 5
    '                y = y + 30

    '                Me.parameterPanel.Controls.Add(a)
    '                Me.parameterPanel.ResumeLayout(False)
    '                Me.ResumeLayout(False)

    '                If rp.ValidValues IsNot Nothing Then
    '                    ''Build listitems
    '                    Dim aList As ArrayList = New ArrayList()
    '                    pvs = rp.ValidValues

    '                    For Each pv As ValidValue In pvs
    '                        aList.Add(New ComboItem(pv.Label, pv.Value))
    '                    Next

    '                    ''Bind listitmes to combobox
    '                    a.DataSource = aList
    '                    a.DisplayMember = "Display"
    '                    a.ValueMember = "Value"
    '                End If
    '            Next

    '        End If



    '    Catch ex As SoapException
    '        MessageBox.Show(ex.Detail.InnerXml.ToString())
    '    End Try



    'End Sub


    'Private Sub buttonOK_Click(sender As System.Object, e As System.EventArgs) Handles buttonOK.Click
    '    parameters = ViewerParameters()
    '    Me.DialogResult = DialogResult.OK
    '    Close()

    'End Sub
    'Private Function ViewerParameters() As Microsoft.Reporting.WinForms.ReportParameter()
    '    Dim numCtrls As Integer = (Me.parameterPanel.Controls.Count / 2)
    '    'Dim rp As List(Of Microsoft.Reporting.WinForms.ReportParameter()) = New List(Of Microsoft.Reporting.WinForms.ReportParameter())
    '    'Dim rp As Microsoft.Reporting.WinForms.ReportParameter() = New Microsoft.Reporting.WinForms.ReportParameter(numCtrls - 1) {}
    '    Dim rp As Microsoft.Reporting.WinForms.ReportParameter() = New Microsoft.Reporting.WinForms.ReportParameter(4) {}

    '    Dim i As Integer = 0

    '    For Each ctrl As Control In Me.parameterPanel.Controls
    '        If TypeOf (ctrl) Is ComboBox Then ' ctrl.GetType = GetType(ComboBox) Then
    '            Dim a As ComboBox = DirectCast(ctrl, ComboBox)
    '            rp(i) = New Microsoft.Reporting.WinForms.ReportParameter()
    '            rp(i).Name = a.Name
    '            If a.SelectedValue IsNot Nothing AndAlso a.SelectedValue.ToString() <> [String].Empty Then
    '                rp(i).Values.Add(a.SelectedValue.ToString())
    '            Else
    '                rp(i).Values.Add(Nothing)
    '            End If
    '            i += 1
    '        End If
    '    Next

    '    Return rp
    'End Function


End Class
Public Class ComboItem
    Public Sub ComboItem(ByVal disp As String, ByVal myvalue As String)
        If disp IsNot Nothing Then
            display = disp
        Else
            display = ""
        End If

        If myvalue IsNot Nothing Then
            value = myvalue
        Else
            value = ""
        End If
    End Sub

    Public display As String
    Public value As String

    Public Overrides Function ToString() As String
        Return display
    End Function

End Class

File: 2012 Book Source Code/Chapter_9/Chapter 9/SSRS Viewer RVC/GetParameters.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Web.Services.Protocols;
using System.Collections;
using SSRS_Viewer_RVC.SSRSWebService;

namespace SSRS_Viewer_RVC
{
	public partial class GetParameters : Form
	{

		private string url;
		private string server;
		private string report;
		private Microsoft.Reporting.WinForms.ReportParameter[] parameters;
		ReportingService2010 rs;

		public GetParameters(string URL)
		{
			InitializeComponent();
			url = URL;
			string[] reportInfo = url.Split('?');
			server = reportInfo[0];
			report = reportInfo[1];
		}

		public Microsoft.Reporting.WinForms.ReportParameter[] Parameters
		{
			get
			{
				return parameters;
			}
		}

		private void GetParameters_Load(object sender, EventArgs e)
		{
			rs = new ReportingService2010();            
			rs.Credentials = System.Net.CredentialCache.DefaultCredentials;

			bool forRendering = true;
			string historyID = null;
			ParameterValue[] values = new ParameterValue[1];
			DataSourceCredentials[] credentials = null;
			ItemParameter[] parametersSSRS = null;            
			ValidValue[] pvs = null;

			int x = 5;
			int y = 30;

			try
			{
                values[0] = new ParameterValue();
                values[0].Label = "ServiceYear";
                values[0].Name = "ServiceYear";
                values[0].Value = "2009";

				parametersSSRS = rs.GetItemParameters(report, historyID, forRendering, values, credentials);                

				if (parametersSSRS != null)
				{
					foreach (ItemParameter rp in parametersSSRS)
					{						
                        this.SuspendLayout();
						this.parameterPanel.SuspendLayout();
						this.parameterPanel.SendToBack();
                        

						// now create a label for the combo box below
						Label lbl = new Label();
						lbl.Anchor = (System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left);
						lbl.Location = new System.Drawing.Point(x, y);
						lbl.Name = rp.Name;
						lbl.Text = rp.Name;
						lbl.Size = new System.Drawing.Size(150, 20);
						this.parameterPanel.Controls.Add(lbl);
						x = x + 150;

						// now make a combo box and fill it
						ComboBox a = new ComboBox();
						a.Anchor = (System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right);
						a.Location = new System.Drawing.Point(x, y);
						a.Name = rp.Name;
						a.Size = new System.Drawing.Size(200, 20);
						x = 5;
						y = y + 30;
						
                        this.parameterPanel.Controls.Add(a);
						this.parameterPanel.ResumeLayout(false);
						this.ResumeLayout(false);

						if (rp.ValidValues != null)
						{
							//Build listitems
							ArrayList aList = new ArrayList();
							pvs = rp.ValidValues;
							foreach (ValidValue pv in pvs)
							{
								aList.Add(new ComboItem(pv.Label, pv.Value));
							}
							//Bind listitmes to combobox
							a.DataSource = aList;
							a.DisplayMember = "Display";
							a.ValueMember = "Value";
						}
					}
				}
			}

			catch (SoapException ex)
			{
				MessageBox.Show(ex.Detail.InnerXml.ToString());
			}
		}

		private Microsoft.Reporting.WinForms.ReportParameter[] ViewerParameters()
		{
			int numCtrls = (this.parameterPanel.Controls.Count / 2);
			Microsoft.Reporting.WinForms.ReportParameter[] rp = new Microsoft.Reporting.WinForms.ReportParameter[numCtrls];
			int i = 0;

			foreach (Control ctrl in this.parameterPanel.Controls)
			{
				if (ctrl.GetType() == typeof(ComboBox))
				{
					ComboBox a = (ComboBox)ctrl;
					rp[i] = new Microsoft.Reporting.WinForms.ReportParameter();
					rp[i].Name = a.Name;
					if (a.SelectedValue != null && a.SelectedValue.ToString() != String.Empty)
					{
						rp[i].Values.Add(a.SelectedValue.ToString());
					}
					else
					{
						rp[i].Values.Add(null);
					}
					i++;
				}
			}

			return rp;
		}

		private void buttonOK_Click(object sender, EventArgs e)
		{
			parameters = ViewerParameters();
			this.DialogResult = DialogResult.OK;
			Close();
		}
	}

	public class ComboItem
	{
		public ComboItem(string disp, string myvalue)
		{
			if (disp != null)
				display = disp;
			else
				display = "";
			if (myvalue != null)
				val = myvalue;
			else
				val = "";
		}

		private string val;
		public string Value
		{
			get { return val; }
			set { val = value; }
		}

		private string display;
		public string Display
		{
			get { return display; }
			set { display = value; }
		}

		public override string ToString()
		{
			return display;
		}
	}
}